projects/packages/password-checker/src/class-password-checker.php

```diff
@@ -0,0 +1,1313 @@
+
+		/**
+		 * Filters the password strength enforcement settings.
+		 *
+		 * You can modify the minimum password length using this filter.
+		 *
+		 * @param int $minimum_password_length minimum password length.
+		 */
+		$this->minimum_password_length = apply_filters( 'password_checker_minimum_password_length', $this->minimum_password_length );
+
+		/**
+		 * Filters the password strength enforcement settings.
+		 *
+		 * You can modify the minimum entropy bits requirement using this filter.
+		 *
+		 * @param int $minimum_entropy_bits minimum entropy bits requirement.
+		 */
+		$this->minimum_entropy_bits = apply_filters( 'password_checker_minimum_entropy_bits', $this->minimum_entropy_bits );
+	}
+
+	/**
+	 * Run tests against a password.
+	 *
+	 * @param string $password      the password.
+	 * @param bool   $required_only only test against required conditions, defaults to false.
+	 *
+	 * @return array an array containing failed and passed test results.
+	 */
+	public function test( $password, $required_only = false ) {
+		// Run the tests.
+		$results = $this->run_tests( $password, $this->get_tests(), $required_only );
+
+		// If we've failed on the required tests, return now.
+		if ( ! empty( $results['failed'] ) ) {
+			return array(
+				'passed'       => false,
+				'test_results' => $results,
+			);
+		}
+
+		$entropy_bits = $this->calculate_entropy_bits( $password );
+
+		// If we have failed the entropy bits test, run the regex tests so we can suggest improvements.
+		if ( $entropy_bits < $this->minimum_entropy_bits ) {
+			$results['failed']['entropy_bits'] = $entropy_bits;
+			// Run the tests.
+			$results = array_merge( $results, $this->run_tests( $password, $this->get_tests( 'preg_match' ) ) );
+		}
+
+		return ( array(
+			'passed'       => empty( $results['failed'] ),
+			'test_results' => $results,
+		) );
+	}
+
+	/**
+	 * Run the tests using the currently set up object values.
+	 *
+	 * @param string $password      the password.
+	 * @param array  $tests         tests to run.
+	 * @param bool   $required_only whether to run only required tests.
+	 *
+	 * @return array test results.
+	 */
+	public function run_tests( $password, $tests, $required_only = false ) {
+		$results = array(
+			'passed' => array(),
+			'failed' => array(),
+		);
+
+		foreach ( $tests as $test_type => $section_tests ) {
+			foreach ( $section_tests as $test_name => $test_data ) {
+				// Skip non-required tests if required_only param is set.
+				if ( $required_only && ! $test_data['required'] ) {
+					continue;
+				}
+
+				$result = call_user_func_array( array( $this, 'test_' . $test_type ), array( $password, $test_data ) );
+				if ( $result ) {
+					$results['passed'][] = array( 'test_name' => $test_name );
+				} else {
+					$results['failed'][] = array(
+						'test_name'   => $test_name,
+						'explanation' => $test_data['error'],
+					);
+
+					if ( isset( $test_data['fail_immediately'] ) ) {
```

Comment:
@briancolinger the fail_immediately property is used to immediately return when that rule fails. To test it we need to create a password that would fail multiple rules but confirm in the output only the fail_immediately results are returned and that the non fail_immediately rule did not run.

Also we would want to have a scenario potentially of 
rule - don't fail_immediately
rule 2 - fail_immediately
rule 3 - don't fail_immediately
and confirm results only show for rule 1 and 2